src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -817,209 +817,238 @@ private BoundCollectionExpression ConvertCollectionExpression(
-            var syntax = node.Syntax;
-            if (LocalRewriter.IsAllocatingRefStructCollectionExpression(node, collectionTypeKind, elementType, Compilation))
+            public BoundCollectionExpression Do()
             {
-                diagnostics.Add(node.HasSpreadElements(out _, out _)
-                    ? ErrorCode.WRN_CollectionExpressionRefStructSpreadMayAllocate
-                    : ErrorCode.WRN_CollectionExpressionRefStructMayAllocate,
-                    syntax, targetType);
-            }
+                var collectionTypeKind = _conversion.GetCollectionExpressionTypeKind(out var elementType, out MethodSymbol? constructor, out bool isExpanded);
 
-            // Stores the `new T(...)` call if this is a normal constructor backed collection (including mutable array
-            // interfaces).  Otherwise stores the `T.Create(...)` call for collection builders.  It will be null for
-            // spans, arrays, and read-only interfaces.
-            BoundExpression? collectionCreation = null;
-            MethodSymbol? collectionBuilderMethod = null;
-            BoundValuePlaceholder? collectionBuilderElementsPlaceholder = null;
+                if (collectionTypeKind == CollectionExpressionTypeKind.None)
+                {
+                    Debug.Assert(_conversion.Kind is ConversionKind.NoConversion);
+                    return _binder.BindCollectionExpressionForErrorRecovery(_node, _targetType, inConversion: false, _diagnostics);
+                }
 
-            switch (collectionTypeKind)
+                // If we're not in the collection-construction case, we do not allow ref-structs.  First, for the
+                // array/interface case, we can't make the storage to store the ref structs in.  Second, for the span
+                // case, spans are not 'allows ref struct' for their T element.  So they don't allow them either.  Finally,
+                // collection builders need to take in a ReadOnlySpan<T> so they are restricted for the same reason.
+                Debug.Assert(elementType is { });
+                if (collectionTypeKind != CollectionExpressionTypeKind.ImplementsIEnumerable &&
+                    elementType.IsRefLikeOrAllowsRefLikeType())
+                {
+                    _diagnostics.Add(ErrorCode.ERR_CollectionRefLikeElementType, _node.Syntax);
+                }
+
+                var result = TryConvertCollectionExpression(collectionTypeKind, elementType, constructor);
+                if (result is null)
+                    return _binder.BindCollectionExpressionForErrorRecovery(_node, _targetType, inConversion: true, _diagnostics);
+
+                result.WasCompilerGenerated = _node.IsParamsArrayOrCollection;
+                result.IsParamsArrayOrCollection = _node.IsParamsArrayOrCollection;
+
+                return result;
+            }
+
+            private readonly BoundCollectionExpression? TryConvertCollectionExpression(
```

Comment:
>TryConvertCollectionExpression [](http://example.com/codeflow?start=56&length=30)

It feels like this should be a local function.

Reply:
i prefer this as a sibling, this is simple linear flow.  I'm fine with how this si structured.  Then when it forks out to many different options, i keep those bound together with locals.